File: python/DesignPatternExamples_python/visitor/visitor_village.py
## @file
#  @brief
#  Implementation of the
#  @ref DesignPatternExamples_python.visitor.visitor_village.Visitor_Village "Visitor_Village"
#  class as used in the @ref visitor_pattern.

from .visitor_class import Visitor
from .visitor_element_classes import *

## Represents a collection of shops that can be visited.
#  
#  This class works as the container of all objects that can be visited.
#  All visits start in this container.
class Visitor_Village:

    ## @name Properties
    #  @{

    ## Property getter for the name of this village: `value = o.Name`
    @property
    def Name(self) -> str:
        return self._name

    ## Property setter for the name of this village: `o.Name = value`
    @Name.setter
    def Name(self, value) -> None:
        self._name = value

    ## @}


    ## Constructor
    def __init__(self) -> None:
        self._shops = [] # type: list[Visitor_Shop]
        self._name = ""
    
    ## @var _shops
    #       List of shops in this village (filled in by the
    #       @ref DesignPatternExamples_python.visitor.visitor_village.Visitor_Village.LoadVillage "LoadVillage()"
    #       method)
    #  @var _name
    #       Name of this village (set by the
    #       @ref DesignPatternExamples_python.visitor.visitor_village.Visitor_Village.LoadVillage "LoadVillage()"
    #       method)

    ## Load the village.  Defines all the shops that can be visited, what each
    #  shop sells, and what each shop needs to make what it sells.
    def LoadVillage(self) -> None:
        self._name = "Village of Self-Sufficiency"

        self._shops.append(Visitor_Restaurant())
        shop = self._shops[-1]
        shop.Name = "Joe's Burger Joint"
        shop.Address = "47 Millings Rd."
        shop.Village = self
        shop.IngredientsForItems = {
                "hamburger": [ "ground beef", "hamburger buns", "ketchup", "mustard", "mayonnaise", "lettuce", "tomato", "onion", "pickles" ]
            }

        self._shops.append(Visitor_Butcher())
        shop = self._shops[-1]
        shop.Name = "Amelia's Butcher Shop"
        shop.Address = "12 Klaxon Ave."
        shop.Village = self
        shop.IngredientsForItems = {
                "ground beef" : []
            }

        self._shops.append(Visitor_Baker())
        shop = self._shops[-1]
        shop.Name = "Oxel's Breads and Buns Bakery"
        shop.Address = "131 Worthington Dr."
        shop.Village = self
        shop.IngredientsForItems = {
                "hamburger buns" : []
            }

        self._shops.append(Visitor_CondimentGrocer())
        shop = self._shops[-1]
        shop.Name = "Connie's Condiments"
        shop.Address = "83 Millings Rd."
        shop.Village = self
        shop.IngredientsForItems = {
                "ketchup" : [ "fresh ketchup" ],
                "mustard" : [ "fresh mustard" ],
                "mayonnaise" : [ "fresh mayonnaise" ]
            }

        self._shops.append(Visitor_VegetableGrocer())
        shop = self._shops[-1]
        shop.Name = "Florence's Vegetables"
        shop.Address = "32 Main St."
        shop.Village = self
        shop.IngredientsForItems = {
                "lettuce" : [],
                "tomato" : [],
                "onion" : [],
                "cucumber" : [],
                "mustard seed" : []
            }

        self._shops.append(Visitor_PickleGrocer())
        shop = self._shops[-1]
        shop.Name = "Larry's Pickle Emporium"
        shop.Address = "34 Main St."
        shop.Village = self
        shop.IngredientsForItems = {
                "pickles" : [ "vinegar", "cucumber", "salt" ]
            }

        self._shops.append(Visitor_Maker())
        shop = self._shops[-1]
        shop.Name = "Klyde and Sons Ketchup Makers"
        shop.Address = "800 Overtown Rd."
        shop.Village = self
        shop.IngredientsForItems = {
                "fresh ketchup" : []
            }

        self._shops.append(Visitor_Maker())
        shop = self._shops[-1]
        shop.Name = "Molly's Mustard Mart"
        shop.Address = "810 Overtown Rd."
        shop.Village = self
        shop.IngredientsForItems = {
                "fresh mustard" : [ "vinegar", "mustard seed" ]
            }

        self._shops.append(Visitor_Maker())
        shop = self._shops[-1]
        shop.Name = "Turk's Mayo Supply"
        shop.Address = "820 Overtown Rd."
        shop.Village = self
        shop.IngredientsForItems = {
                "fresh mayonnaise" : []
            }

        self._shops.append(Visitor_Maker())
        shop = self._shops[-1]
        shop.Name = "Vinnies' Sour Flavors"
        shop.Address = "830 Overtown Rd."
        shop.Village = self
        shop.IngredientsForItems = {
                "vinegar" : []
            }

        self._shops.append(Visitor_Maker())
        shop = self._shops[-1]
        shop.Name = "Jessie's Salt Works"
        shop.Address = "920 Overtown Rd."
        shop.Village = self
        shop.IngredientsForItems = {
                "salt" : []
            }

    ## Accept a visitor and send them around to all the shops.
    def Accept(self, visitor : Visitor) -> None:
        for shop in self._shops:
            shop.Accept(visitor)

File: rust/design_pattern_examples_rust/src/visitor/visitor_village.rs
//! Contains the Village struct that holds all the shops that can be visited.

//-----------------------------------------------------------------------------
use std::collections::HashMap;

use super::visitor_shop::VisitorShop;
use super::visitor_ordervisitor::OrderVisitor;

//-----------------------------------------------------------------------------

/// Represents a collection of shops that can be visited.
pub struct Village {
    /// Name of this village.
    pub name: String,
    /// List of shops in this village.
    shops: Vec<VisitorShop>,
}

impl Village {
    /// Constructor
    ///
    /// # Returns
    /// Returns a new instance of the Village struct.  Call load() to load
    /// the particulars of this instance.
    pub fn new() -> Village {
        Village {
            name: String::new(),
            shops: vec![],
        }
    }

    /// Set up the specified Village object with all the shops that can be
    /// visited.
    pub fn load(&mut self) {
        self.name = String::from("Village of Self-Sufficiency");

        let mut shop_ingredients: HashMap<String, Vec<String>> = HashMap::new();
        shop_ingredients.insert("hamburger".to_string(), vec!["ground beef", "hamburger buns", "ketchup", "mustard", "mayonnaise", "lettuce", "tomato", "onion", "pickles"].iter().map(|x|{x.to_string()}).collect());
        let mut shop = VisitorShop::new(
            "Joe's Burger Joint",
            "47 Millings Rd.",
            shop_ingredients
            );
        self.shops.push(shop);

        shop_ingredients = HashMap::new();
        shop_ingredients.insert("ground beef".to_string(), vec![]);
        shop = VisitorShop::new(
            "Amelia's Butcher Shop",
            "12 Klaxon Ave.",
            shop_ingredients,
        );
        self.shops.push(shop);

        shop_ingredients = HashMap::new();
        shop_ingredients.insert("hamburger buns".to_string(), vec![]);
        shop = VisitorShop::new(
            "Oxel's Breads and Buns Bakery",
            "131 Worthington Dr.",
            shop_ingredients,
        );
        self.shops.push(shop);

        shop_ingredients = HashMap::new();
        shop_ingredients.insert("ketchup".to_string(), vec!["fresh ketchup"].iter().map(|x|{x.to_string()}).collect());
        shop_ingredients.insert("mustard".to_string(), vec!["fresh mustard"].iter().map(|x|{x.to_string()}).collect());
        shop_ingredients.insert("mayonnaise".to_string(), vec!["fresh mayonnaise"].iter().map(|x|{x.to_string()}).collect());
        shop = VisitorShop::new(
            "Connie's Condiments",
            "83 Millings Rd.",
            shop_ingredients,
        );
        self.shops.push(shop);

        shop_ingredients = HashMap::new();
        shop_ingredients.insert("lettuce".to_string(), vec![]);
        shop_ingredients.insert("tomato".to_string(), vec![]);
        shop_ingredients.insert("onion".to_string(), vec![]);
        shop_ingredients.insert("cucumber".to_string(), vec![]);
        shop_ingredients.insert("mustard seed".to_string(), vec![]);
        shop = VisitorShop::new(
            "Florence's Vegetables",
            "32 Main St.",
            shop_ingredients,
        );
        self.shops.push(shop);

        shop_ingredients = HashMap::new();
        shop_ingredients.insert("pickles".to_string(), vec!["vinegar", "cucumber", "salt"].iter().map(|x|{x.to_string()}).collect());
        shop = VisitorShop::new(
            "Larry's Pickle Emporium",
            "34 Main St.",
            shop_ingredients,
        );
        self.shops.push(shop);

        shop_ingredients = HashMap::new();
        shop_ingredients.insert("fresh ketchup".to_string(), vec![]);
        shop = VisitorShop::new(
            "Klyde and Sons Ketchup Makers",
            "800 Overtown Rd.",
            shop_ingredients,
        );
        self.shops.push(shop);

        shop_ingredients = HashMap::new();
        shop_ingredients.insert("fresh mustard".to_string(), vec!["vinegar", "mustard seed"].iter().map(|x|{x.to_string()}).collect());
        shop = VisitorShop::new(
            "Molly's Mustard Mart",
            "810 Overtown Rd.",
            shop_ingredients,
        );
        self.shops.push(shop);

        shop_ingredients = HashMap::new();
        shop_ingredients.insert("fresh mayonnaise".to_string(), vec![]);
        shop = VisitorShop::new(
            "Turk's Mayo Supply",
            "820 Overtown Rd.",
            shop_ingredients,
        );
        self.shops.push(shop);

        shop_ingredients = HashMap::new();
        shop_ingredients.insert("vinegar".to_string(), vec![]);
        shop = VisitorShop::new(
            "Vinnies' Sour Flavors",
            "830 Overtown Rd.",
            shop_ingredients,
        );
        self.shops.push(shop);

        shop_ingredients = HashMap::new();
        shop_ingredients.insert("salt".to_string(), vec![]);
        shop = VisitorShop::new(
            "Jessie's Salt Works",
            "920 Overtown Rd.",
            shop_ingredients,
        );
        self.shops.push(shop);
    }

    /// Visit all shops in the given Village object to find the ingredients
    /// specified in the OrderVisitor object.  Can set off a nested chain of
    /// calls.
    ///
    /// # Parameters
    /// - order
    ///
    ///   An OrderVisitor object that contains a list of items to order and
    ///   a list of items received after the visit is complete.
    pub fn visit(&mut self, order: &mut OrderVisitor) {
        // Clone the shop list from the village to get around an error
        // involving two mutable references to the village: The first is in
        // the for loop and the second is when passing the village to the
        // visit_shop() method.  Cloning the shop is kind of expensive,
        // especially since this method can be recursively called by shops as
        // needed.  But I could not find another way to get around the double
        // mutable error.
        let mut shops = self.shops.clone();

        for shop in shops.iter_mut() {
            order.visit_shop(shop, self);
        }
    }
}
